mem_windows.go

```diff
@@ -25,27 +25,43 @@ func (ca *Cagent) MemResults() (MeasurementsMap, error) {
 
 	memStat, err := mem.VirtualMemoryWithContext(ctx)
 
-	results["free_B"] = nil
-	results["shared_B"] = nil
-	results["buff_B"] = nil
+	results = map[string]interface{}{
+		"total_B":           nil,
+		"free_B":            nil,
+		"free_percent":      nil,
+		"cached_B":          nil,
+		"cached_percent":    nil,
+		"used_B":            nil,
+		"used_percent":      nil,
+		"available_B":       nil,
+		"available_percent": nil,
+	}
 
 	if err != nil {
 		log.Errorf("[MEM] Failed to get virtual memory stat: %s", err.Error())
 		errs = append(errs, err.Error())
-		results["total_B"] = nil
-		results["used_B"] = nil
-		results["available_B"] = nil
 	} else {
 		results["total_B"] = memStat.Total
-		results["used_B"] = memStat.Used
 		results["available_B"] = memStat.Available
+		results["available_percent"] = floatToIntPercent((float64(memStat.Available)) / float64(memStat.Total))
 	}
 
 	free, err := watcher.Query(`\Memory\Free & Zero Page List Bytes`, "*")
 	if err != nil {
 		log.Errorf("[MEM] Failed to get free memory: %s", err.Error())
 	} else {
-		results["free_B"] = free
+		results["used_B"] = int(memStat.Total) - int(free)
+		results["used_percent"] = floatToIntPercent((float64(memStat.Total) - free) / float64(memStat.Total))
+		results["free_B"] = int(free)
+		results["free_percent"] = floatToIntPercent(free / float64(memStat.Total))
+	}
+
+	cached, err := watcher.Query(`\Memory\Cache Bytes`, "*")
+	if err != nil {
+		log.Errorf("[MEM] Failed to get cached memory: %s", err.Error())
+	} else {
+		results["cached_B"] = int(cached)
+		results["cached_percent"] = floatToIntPercent(float64(cached) / float64(memStat.Total))
 	}
 
 	if len(errs) == 0 {
```

Comment:
Same as before, it seems there can only occur one error in the function.
Also as a good practice:
Try to invert your flow. You check for the happy path and expect a failure in the end. Inverting this makes the code much easier to read.
In the end of the function you should have the successful return statement. Inside the function try to check for the error cases and then immediately return (if possible).

Reply:
actually, it could be more than 1 error. Fixed

Reply:
>Try to invert your flow
I do this approach when possible. In this case, I needed to add multiple errors but still continue. I've just forgotten to append some of them 😬